### PR TITLE
Use the gain for full cbf files 

### PR DESCRIPTION
**Pull request by phyy-nx**
- Add a test for dxtbx.image_average that includes testing gain
- Modify full cbf format readers to read the gain
- Modify cbf_writer to use the gain

This does affect any processing that uses cbfs with a gain !=1.  See for example dials_regression/image_examples/SPring8_ADSC_SN916/Xtal17-2phi_3_015.cbf.  On master, dxtbx.print_header will report gain=1.  With this change, gain will be 1.84.

### CODE REVIEW

File: format/FormatCBFMultiTile.py [hunk 1]
```diff
@@ -126,12 +126,20 @@ def _detector(self):
...
+                overload = cbf.get_overload(i)
                 trusted_range = (underload, overload)
             except Exception:
                 trusted_range = (0.0, 0.0)
 
+            try:
+                cbf.find_column(b"gain")
+                cbf.select_row(i)
+                gain = cbf.get_doublevalue()
+            except Exception:
```
Comment (Anthchirp) [inline format/FormatCBFMultiTile.py]: This introduces 3 `except Exception:`. Are there more specific exceptions we could use?

File: model/detector_helpers.py [hunk 2]
```diff
@@ -115,6 +115,17 @@ def find_undefined_value(cbf_handle):
     return cbf_handle.get_doublevalue()
 
 
+def find_gain_value(cbf_handle):
+    """Given a cbf handle, get the gain value."""
+    try:
+        cbf_handle.find_category(b"array_intensities")
+        cbf_handle.find_column(b"gain")
+    except Exception as e:
+        assert "CBF_NOTFOUND" in str(e), str(e)
```
Comment (Anthchirp) [inline model/detector_helpers.py]:
```suggestion
        if "CBF_NOTFOUND" not in str(e):
            raise
```

File: format/FormatCBFMultiTile.py (hunk 1 above)
Comment (phyy-nx) [reply format/FormatCBFMultiTile.py]: No, but I've brought in a string check

Comment (Anthchirp) [reply format/FormatCBFMultiTile.py]: :+1: thanks

**Comment by phyy-nx** [comment]
Additional bug found and squashed.  Ready to review/re-review.  Thanks.

**Comment by phyy-nx** [comment]
MultiTileHierarchy: image_examples/LCLS_cspad_nexus/idx-20130301060858401.cbf
CBFFull: image_examples/SPring8_ADSC_SN916/Xtal17-2phi_3_015.cbf

**Comment by graeme-winter** [comment]
@phyy-nx I was thinking of ones you _would not_ have already tested with 😉 is fine tho, got some here.

**Comment by phyy-nx** [comment]
Splitting out the fixes would be tricky, there are already 4 commits, one of which is fixing problems with cbf sections.  I can do if needed tho.

**Comment by graeme-winter** [comment]
Not too hung up on it, but since I was asked to review I thought it would be part of what I would ask for 🙂

OK, I have tested with "normal" CBF files which are full from Diamond and this appears to _do no harm_ i.e. the gain is supposed to be 1 and this is retained across the change set

**Comment by phyy-nx** [comment]
Re-wrote the commits to separate bugfixes from gain work

**Pull request merged**